### PR TITLE
#23: normalizing Scala -> PG objects

### PR DESCRIPTION
**Pull request by lsulak**
Related to #23, but won't fully implement everything in that ticket. I'll split it into more PRs because #23 is dependant on DB code that is still not fully done&merged.

### CODE REVIEW

File: server/src/main/scala/za/co/absa/atum/web/api/database/Runs.scala [hunk 1]
```diff
@@ -34,32 +36,37 @@ class Runs (implicit dBEngine: SlickPgEngine) extends DBSchema{
...
-      .replace("[", "{")
-      .replace("]", "}")
+  private def nestedScalaSeqToPgArray(toConvert: Seq[Seq[String]]): String = {
+    val scalaSeqJsonized = toConvert.map(scalaSeqToPgArray).mkString(",")
+    s"{$scalaSeqJsonized}"
+  }
+
+  private def scalaSeqToPgArray(toConvert: Seq[String]): String = {
+    val scalaArrayAndItsContent: Regex = "^\\[(.*)\\]$".r
+    val scalaSeqJsonized = SerializationUtils.asJson(toConvert)
```
[the remark this answers is not in the document]
Comment (lsulak) [reply server/src/main/scala/za/co/absa/atum/web/api/database/Runs.scala]: Explained via a quick chat

File: server/src/main/scala/za/co/absa/atum/server/api/database/Runs.scala [hunk 2]
```diff
@@ -35,33 +37,34 @@ class Runs (implicit dBEngine: SlickPgEngine) extends DBSchema{
...
-      .replace("[", "{")
-      .replace("]", "}")
+  private def nestedScalaSeqToPgArray[T <: AnyRef: ClassTag](toConvert: Seq[Seq[T]]): String = {
+    val scalaSeqJsonized = toConvert.map(scalaSeqToPgArray).mkString(",")
+    "{" + scalaSeqJsonized + "}"
+  }
+
+  private def scalaSeqToPgArray[T <: AnyRef: ClassTag](toConvert: Seq[T]): String = {
+    val scalaSeqJsonized = SerializationUtils.asJson(toConvert)  // this also correctly escapes double quotes
+    "{" + scalaSeqJsonized.substring(1, scalaSeqJsonized.length - 1) + "}"
```
Comment (benedeki) [inline server/src/main/scala/za/co/absa/atum/server/api/database/Runs.scala:47]: **Minor:**
```suggestion
    val scalaSeqJsonized = SerializationUtils.asJson(toConvert)                  // this also correctly escapes double quotes
    "{" + scalaSeqJsonized.substring(1, scalaSeqJsonized.length - 1) + "}"  //...only the square brackets have to be replaced with curly ones
```

File: server/src/main/scala/za/co/absa/atum/server/api/database/Runs.scala [hunk 3]
```diff
@@ -35,33 +37,34 @@ class Runs (implicit dBEngine: SlickPgEngine) extends DBSchema{
...
-  private def scalaIterableToSQLArray(toConvert: Seq[Any]): String = {
-    SerializationUtils.asJson(toConvert)
-      .replace("[", "{")
-      .replace("]", "}")
+  private def nestedScalaSeqToPgArray[T <: AnyRef: ClassTag](toConvert: Seq[Seq[T]]): String = {
+    val scalaSeqJsonized = toConvert.map(scalaSeqToPgArray).mkString(",")
+    "{" + scalaSeqJsonized + "}"
+  }
+
+  private def scalaSeqToPgArray[T <: AnyRef: ClassTag](toConvert: Seq[T]): String = {
```
Comment (benedeki) [inline server/src/main/scala/za/co/absa/atum/server/api/database/Runs.scala:45]: Why the type limitation (`<: AnyRef: ClassTag`)?

**Review by benedeki — APPROVED** [review]
Maybe there's a generic recursive solution, but this is good and enough:

- [x] code reviewed
- [ ] pulled
- [ ] built
- [ ] run

File: server/src/main/scala/za/co/absa/atum/server/api/database/Runs.scala (hunk 3 above)
Comment (lsulak) [reply server/src/main/scala/za/co/absa/atum/server/api/database/Runs.scala:45]: This is the definition of `asJson` function that I'm calling inside: `def asJson[T <: AnyRef](obj: T): String` and this generic type `T` here is also limited, so the 'outer' `T` here in my function cannot be a superset of it. The build was failing

File: server/src/main/scala/za/co/absa/atum/server/api/database/Runs.scala (hunk 2 above)
Comment (lsulak) [reply server/src/main/scala/za/co/absa/atum/server/api/database/Runs.scala:47]: That's better, thanks, I'll change it :)

**Review by benedeki — APPROVED** [review]

**Pull request merged**